app/grandchallenge/charts/static/js/charts/render_charts.mjs

```diff
@@ -41,9 +34,33 @@ const observer = new IntersectionObserver(handleInterSection, {
   threshold: 1.0,
 });
 
-
 document.addEventListener("DOMContentLoaded", function(event) {
-    for (const element of document.getElementsByClassName("vega-lite-chart-lazy")) {
+    for (const element of document.getElementsByClassName("vega-lite-chart")) {
         observer.observe(element, );
     }
 });
+
+
+function mutationObserverCallback(mutationList, observer) {
+    mutationList.forEach((mutation) => {
+        mutation.addedNodes.forEach((addedNode) => {
+            if (addedNode.nodeType !== Node.TEXT_NODE) {
+                for (const element of addedNode.getElementsByClassName("vega-lite-chart")) {
+                    if (!element.getAttribute(handledAttribute)) {
+                        renderVegaLiteChart(element);
+                        element.setAttribute(handledAttribute, "");
+                    }
+                }
+            }
+        })
+    })
+}
+
+const mutationObserver = new MutationObserver(mutationObserverCallback);
+
+if (document.getElementById("ajaxDataTable")) {
```

Comment:
`render_charts.js` now needs to know about ajax data tables. This also won't work if the table has a different id, or there are multiple tables in the DOM. Can't the mutationObserver be exported and the tables that contain charts be registered?

Reply:
Since the render_charts.mjs is loaded on the page as a JS module, the variables defined inside it have a module scope and not global scope, so they are not accessible from other scripts. That was the reason I asked the question on Teams chat since I was trying to make the Datatable register itself to the observer and I am not sure if render_charts.mjs is actually a module since its logic now need to know about things happening in the page and not entirely self-contained, maybe I am wrong and I need more reading on this topic.

One suggestion is to make the observer observe the body and not specifically to an element so it will detect any chart whenever added to the screen. Also, since the view are not heavily dynamic (i.e. with many JS node additions/deletions), I don't expect that to affect performance and the observer only acts on node additions and on a specific class. What do you think?

Reply:
Thanks! I am implementing it. A new thing I learned which is that `import` statement can not be used outside a module. So the file `list.js` should be loaded as a module in the page. I am testing it now.

Reply:
Yes, in general we should refactor JS files to modules, but each one needs to be checked carefully as it doesn't always work.